### ISSUE TITLE
Elixir project creation

### ISSUE THREAD

**Issue by KronicDeth** [KronicDeth/intellij-elixir#22]
Allow the creation of Elixir projects from the New Project menu.  This will include the use of mix to create the new project template. Feature request by @Claudiojlf in https://github.com/KronicDeth/intellij-elixir/issues/6#issuecomment-59466280.  Requested by @prodev42 in #77 and @andreacfm on [stackoverflow.com](https://stackoverflow.com/questions/27730087/new-elixir-project-in-intellij#comment43899731_27730087)
# References
- https://github.com/ignatov/intellij-erlang/commit/28e4f1e650fafc0156f66a9e4981668442096e05
- https://github.com/ignatov/intellij-erlang/commit/b9d2eda1618b4e352ff239005b857c1bdeb4ac98
# Tasks
- [x] Setup SDK (#79)
- [x] Add moduleType for Elixir Project
- [ ] Allow setting of `mix new` options using mix from Setup SDK.

**Comment by KronicDeth** [comment]
From looking at @ignatov's [work on intellij-erlang](https://github.com/ignatov/intellij-erlang/commit/28e4f1e650fafc0156f66a9e4981668442096e05), I think New Project uses moduleTypes for the list of allowed languages.  Further research is needed.

**Comment by zyuyou** [comment]
hello, I have rewrite the sdkType part which you have implemented, and the new one is more like intellij-erlang.
![new project](https://cloud.githubusercontent.com/assets/1913492/8613951/5cdafdbe-2717-11e5-81c5-e0ecd9dc44c3.png)
<new Project> is very simple, but it will create the source-dir **lib** instead of **src**.
![mix-importwizard-1](https://cloud.githubusercontent.com/assets/1913492/8614277/af5abfea-271a-11e5-8026-817c519add15.png)
![mix-importwizard-2](https://cloud.githubusercontent.com/assets/1913492/8614278/af61e11c-271a-11e5-9887-fd6859733658.png)
![projectdetector](https://cloud.githubusercontent.com/assets/1913492/8614324/2a4ffd78-271b-11e5-8449-58956291e264.png)
and I also add importWizard using **mix** , new ProjectDetector and mix runconfig.

in the end, I think the lask task: "Allow setting of mix new options using mix from Setup SDK." would more like the android-plugin's <new Project> , and I will working on it.
![mix new feature](https://cloud.githubusercontent.com/assets/1913492/8614418/0bc35f02-271c-11e5-839a-3ef9bc237f62.png)

**Comment by KronicDeth** [comment]
Very nice @zyuyou, are you planning on opening a pull request now or waiting until you finish the "Allow setting of mix new options using mix from Setup SDK"?  Being able to run `mix compile` and `mix clean` is already a huge feature.  More PRs that are tightly focused are good because (1) it's easier for me to code review them; (2) it's easier to describe them in the changelog; and (3) if you have more PRs you'll get more mentions of your github handle in the changelog for more open source glory. :grin:

**Comment by zyuyou** [comment]
yes, and I will take time to do that，because it need some test-case and be tested in other idea products.
one more thing is what are **idea.xml**, **intellij-elixir.xml**, **intellij_elixir.xml** , **module_intellij-elixir.xml** the four files for? and how to use it? Although, I known that some for get dev-env and some for package the product-plugin.

**Comment by KronicDeth** [comment]
The `*.xml` files are used to run [the build on travis-ci.org](https://travis-ci.org/KronicDeth/intellij-elixir) [using ant](https://github.com/KronicDeth/intellij-elixir/blob/0e9300c94da9a981e39a540740937e48dc27cead/.travis.yml#L9-L10) because the [run configuration](https://github.com/KronicDeth/intellij-elixir/tree/0e9300c94da9a981e39a540740937e48dc27cead/.idea/runConfigurations) in the IDEA project, such as the [`org.elixir_lang` test run configuration](https://github.com/KronicDeth/intellij-elixir/blob/0e9300c94da9a981e39a540740937e48dc27cead/.idea/runConfigurations/org_elixir_lang.xml) cannot be run on travis-ci.org.

As far as I was been able to research those run configurations would only work on a [Team City](https://www.jetbrains.com/teamcity/) server and there was no easily accessible, free for open source, Team City service when I setup the build.  Also, I use TravisCI at work, so I understood how to set it up.

**Comment by zyuyou** [comment]
ok, thanks!

**Comment by KronicDeth** [comment]
This will be resolved by partially by #167

**Comment by sh41** [comment]
All three tasks in this issue are done.

- **Setup SDK** and **moduleType for Elixir Project** were already checked off here.
- **"Allow setting of `mix new` options using mix from Setup SDK"** shipped in
  [v13.1.0](https://github.com/KronicDeth/intellij-elixir/releases/tag/v13.1.0)
  ([780ab3da9](https://github.com/KronicDeth/intellij-elixir/commit/780ab3da9f24dca8258f0e733e15194037afdd65)).
  New Project → Elixir now has a `mix new Options` group with `--app`, `--module`, `--sup` and
  `--umbrella` — every option `mix new` takes — and runs `mix` from the Elixir SDK selected in the
  wizard itself, which is what "using mix from Setup SDK" asked for.

Reviewing this issue turned up several problems in what the wizard configured *after* running
`mix new`, all fixed in [#3919](https://github.com/KronicDeth/intellij-elixir/pull/3919):

- the module's compiler output pointed at a directory that never existed whenever `--app` was left
  blank, which is the default;
- ticking `--sup` alongside `--umbrella` appeared to work and did nothing, because `mix new` ignores
  it at an umbrella root without reporting anything;
- umbrella projects were given a source path and a compiler output that cannot apply to a root with
  no application of its own, one of which left a stray empty directory behind;
- an app added under an umbrella's `apps/` never got its `lib/` and `test/` marked, and nothing
  reported it;
- every project the plugin created was reported as having an outdated format the next time it was
  opened.

One gap worth naming, since it isn't one of this issue's tasks: the New Project entry is registered
in the plugin's Java-dependent descriptor, so it doesn't appear in WebStorm, PyCharm or RubyMine.
That's tracked by [#3686](https://github.com/KronicDeth/intellij-elixir/issues/3686).

One larger question is out of this issue's scope and now has its own:
[#3920](https://github.com/KronicDeth/intellij-elixir/issues/3920). An umbrella created here has no
child applications, so the umbrella option gives you a correct but empty shell, and the wizard treats
`--umbrella` as a fourth `mix new` option when it actually selects a different kind of project.

Closing this as completed.

---

Closed as part of an AI-assisted review of the backlog. A human checked this one before closing it,
but it might still be a mistake. If you think the reason above is wrong, please reopen the issue and
tell us why.
